### PR TITLE
[clang-tidy] Add bsl-special-member-functions

### PR DESCRIPTION
**Pull request by ytuo**
A12-0-1
Checks for classes where some, but not all, of the special member functions are defined
Modified existing special-member-functions (added constructor)

### CODE REVIEW

File: clang-tools-extra/test/clang-tidy/checkers/bsl-special-member-functions.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+};
+// CHECK-MESSAGES: [[@LINE-3]]:7: warning: class 'DefinesConstructor' defines a default constructor but does not define a destructor, a copy constructor, a copy assignment operator, a move constructor or a move assignment operator [bsl-special-member-functions]
+
+class DefinesDestructor {
+  ~DefinesDestructor();
+};
+// CHECK-MESSAGES: [[@LINE-3]]:7: warning: class 'DefinesDestructor' defines a non-default destructor but does not define a copy constructor, a copy assignment operator, a move constructor or a move assignment operator [bsl-special-member-functions]
+
+class DefinesDefaultedDestructor {
+  ~DefinesDefaultedDestructor() = default;
```
Comment (rianquinn) [inline clang-tools-extra/test/clang-tidy/checkers/bsl-special-member-functions.cpp:19]: Need to add a test for virtual. Specifically, you should be allowed to define a pure virtual, or default virtual destructor without this test triggering as that is the only way to define a virtual base.

File: clang-tools-extra/test/clang-tidy/checkers/bsl-special-member-functions.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+};
+// CHECK-MESSAGES: [[@LINE-3]]:7: warning: class 'DefinesDefaultedDestructor' defines a default destructor but does not define a copy constructor, a copy assignment operator, a move constructor or a move assignment operator [bsl-special-member-functions]
+
+class DefinesCopyConstructor {
+  DefinesCopyConstructor(const DefinesCopyConstructor &);
```
Comment (rianquinn) [inline clang-tools-extra/test/clang-tidy/checkers/bsl-special-member-functions.cpp:32]: Need default versions of these as well. Basically, if you add define a default, you need to define them all.

File: clang-tools-extra/test/clang-tidy/checkers/bsl-special-member-functions.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
+// RUN: %check_clang_tidy %s bsl-special-member-functions %t
+
+class DefinesConstructor {
+  DefinesConstructor() = default;
```
Comment (rianquinn) [inline clang-tools-extra/test/clang-tidy/checkers/bsl-special-member-functions.cpp]: Need to add a test for non-default constructors as well as when you delete the default constructor.

**Comment by ytuo** [comment]
ready (?)

**Pull request merged**